File: research/explorations/bench-realistic/nested-calls.py
import sys, time

def addn(a, b):
    return a + b

def muln(a, b):
    return a * b

def compute(x, y):
    a = muln(x, y)
    b = addn(a, x)
    return addn(b, y)

def bench(n):
    s = 0
    i = 0
    while i < n:
        j = i + 1
        s += compute(i, j)
        i += 1
    return s

N = int(sys.argv[1]) if len(sys.argv) > 1 else 1000
# warmup
for _ in range(100):
    bench(N)
iters = 10000
start = time.monotonic_ns()
for _ in range(iters):
    r = bench(N)
elapsed = time.monotonic_ns() - start
print(f"result:     {r}")
print(f"iterations: {iters}")
print(f"total:      {elapsed / 1e6:.2f}ms")
print(f"per call:   {elapsed // iters}ns")

File: research/explorations/bench-realistic/nested-calls.js
function addn(a, b) { return a + b; }
function muln(a, b) { return a * b; }

function compute(x, y) {
    const a = muln(x, y);
    const b = addn(a, x);
    return addn(b, y);
}

function bench(n) {
    let s = 0;
    for (let i = 0; i < n; i++) {
        s += compute(i, i + 1);
    }
    return s;
}

const N = parseInt(process.argv[2]) || 1000;
// warmup
for (let i = 0; i < 1000; i++) bench(N);
const iters = 10000;
const start = process.hrtime.bigint();
let r;
for (let i = 0; i < iters; i++) r = bench(N);
const elapsed = Number(process.hrtime.bigint() - start);
console.log(`result:     ${r}`);
console.log(`iterations: ${iters}`);
console.log(`total:      ${(elapsed / 1e6).toFixed(2)}ms`);
console.log(`per call:   ${Math.floor(elapsed / iters)}ns`);
